File: day1.py
# set up -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# load data --------------------------------------------------------------------
spend_report = pd.read_csv('day1.csv', header = None)  
spend_report = spend_report.rename(columns = {0:'cost'})

# solve the puzzle -------------------------------------------------------------
# Part 1 (which 2 numbers sum to 2020 -- multiply these together)
stop = False 

for i in range(len(spend_report)):
    for j in range(len(spend_report)):
        if(spend_report['cost'][i] + spend_report['cost'][j] == 2020):
            print(spend_report['cost'][i] * spend_report['cost'][j])
            stop = True
            break
        if (stop):
            break

# Part 2 (which 3 numbers sum to 2020 -- multiply these together)
stop = False 

for i in range(len(spend_report)):
    for j in range(len(spend_report)):
        if(spend_report['cost'][i] + spend_report['cost'][j] <= 2020):
            for k in range(len(spend_report)):
                if(spend_report['cost'][i] + spend_report['cost'][j] + spend_report['cost'][k] == 2020):
                    print(spend_report['cost'][i] * spend_report['cost'][j] * spend_report['cost'][k])
                    stop = True
                    break
            if (stop):
                break
        if (stop):
            break

    
# An alternative neat solution found online (@ASpittal) -------------------------------------------

def part_one(nums, set_nums):
    for num in nums:
        if 2020 - num in set_nums:
            return num * (2020 - num)


def part_two(nums, set_nums):
    for num in nums:
        for num2 in nums:
            if (2020 - num - num2) in set_nums:
                return num * (2020 - num - num2) * num2



nums = [int(line) for line in spend_report['cost']]
set_nums = set(nums)

print("Part 1", part_one(nums, set_nums))
print("Part 2", part_two(nums, set_nums))
    

File: day1.R
# Decription of script ---------------------------------------------------------
# Day 1 advent of code -- adding and multiplying numbers.
# ------------------------------------------------------------------------------

# set up -----------------------------------------------------------------------
library(tidyverse)

# load data --------------------------------------------------------------------
spend_report <- read_csv(paste0(here::here(), "/day1.csv"), col_names = FALSE)
spend_report <- spend_report %>%
  rename(cost = X1)

# solve the puzzle -------------------------------------------------------------
# Part 1 (which 2 numbers sum to 2020 -- multiply these together)
for (i in 1:nrow(spend_report)){
  adds_to_2020 <- which((spend_report$cost[i] + spend_report$cost == 2020)) # add all the other costs to cost i and find which (if any) is equal to 2020
  if(length(adds_to_2020) > 0){ # if any do add to 2020 then multiply together and stop the loop
    print(spend_report$cost[i] * spend_report$cost[adds_to_2020])
    break
  }
}

# Part 2 (which 3 numbers sum to 2020 -- multiply these together)
stop <- FALSE
for (i in 1:nrow(spend_report)){
  for (j in 1:nrow(spend_report)){
    adds_to_2020 <- which((spend_report$cost[i] + spend_report$cost[j] + spend_report$cost == 2020)) # add all the other costs to cost i and find which (if any) is equal to 2020
    if(length(adds_to_2020) > 0){ # if any do add to 2020 then multiply together and stop the loop
      print(spend_report$cost[i] * spend_report$cost[j] * spend_report$cost[adds_to_2020])
      print(paste(i, j, adds_to_2020))
      stop = TRUE
      break
    }
    if(stop){
      break
    }
  }
}
